index.rst

```diff
@@ -1,61 +1,173 @@
+
+ - DMS-PRTL-REQ-0033 "Queries on the Alerts Database" :cite:`LDM-554`:
+
+     The Portal aspect shall provide a query interface to the Alert Database, allowing searches based on parameters which shall include, but may not be limited to, Alert ID, time of alert, position on the sky, filter, and alert characteristics.
+
+ - DMS-PRTL-REQ-0048 "Alert Visualization" :cite:`LDM-554`:
+
+     The Portal aspect shall provide for the users a "property sheet" for the contents of an alert packet including, but not necessarily limited to, the alert postage stamp image, the postage stamp time series, the photometric time series, the source and object information (e.g., position, brightness).
+
+Proposed Implementation
+=======================
+
+We can satisfy these design inputs by storing serialized Avro alert data (the same bytes sent via Kafka to brokers) in a S3-like object store, indexed by a unique alert ID.
+Each alert packet corresponds to one object in the object store.
 
 .. note::
 
-   **This technote is not yet published.**
+   An alternative would be to combine many packets into a block in the object store, perhaps of about 100 alert packets.
+   This might permit more efficient storage and retrieval.
+   Storage might be more efficient because compression would be better when storing many alerts.
+   Retrieval might be more efficient because it might tune the outgoing flows into a smaller number of TCP connections which get a chance to grow window sizes.
 
-   Design document for a database of all alerts sent by the Rubin observatory.
+   But this would be more complex, and make writing more difficult, as writes need to append to existing data which would require coordination between writers, so this design sticks to a simpler structure.
+
+An object store is used because it is cheap, scales well to handle terabytes of data, and should support parallel retrieval reasonably well.
+
+Writing data
+------------
+
+When the alert production pipeline has computed a new alert packet, it should be careful to write it to the alert archive before publishing to a Kafka topic, to ensure that it gets archived. The overall flow should be:
+
+ 1. Compute the alert packet payload, including generating a unique alert ID which should eventually appear in the PPDB.
+ 2. Write the alert packet to the object store, using the alert ID as a key.
+ 3. Publish the alert packet to the Kafka topic that serves data to community brokers.
+
+In addition, after all observing is complete for a night, all the alerts that were succesfully published that night should be combined into a single Avro Object Container file.
+The file should be stored on an archival filesystem.
+The set of published alerts can be identified by consuming from the Kafka topic.
+
+
+Reading data
+------------
+
+To read individual alert data, users access the backing alert packets through the butler, which should wrap up the object storage and provide access by alert ID.
```

Comment:
I think of the Hybrid Alert system as a separate API layered on top of this one. We can present that datalink-type API separately, since it's easy to translate into the alert DB queries.

That Hybrid Alert API has other needs anyway, like rate limiting.